src/java_bytecode/select_pointer_type.cpp

```diff
@@ -11,15 +11,101 @@
 /// classes to concrete versions).
 
 #include "select_pointer_type.h"
+#include "java_types.h"
 #include <util/std_types.h>
+#include "java_object_factory.h"
 
 /// Select what type should be used for a given pointer type. For the base class
 /// we just use the supplied type. Derived classes can override this behaviour
 /// to provide more sophisticated type selection
 /// \param pointer_type: The pointer type replace
 /// \return A pointer type where the subtype may have been modified
 pointer_typet select_pointer_typet::convert_pointer_type(
-  const pointer_typet &pointer_type, const namespacet &ns) const
+  const pointer_typet &pointer_type,
+  const generic_recursion_mapt &generic_recursion_map,
+  const namespacet &ns) const
 {
-  return  pointer_type;
+  // if we have a map of generic parameters -> types and the pointer is
+  // a generic parameter or type, specialize the pointer with concrete types
+  if(
+    !generic_recursion_map.empty() &&
+    (is_java_generic_parameter(pointer_type) ||
+     is_java_generic_type(pointer_type)))
+  {
+    return specialize_generics(pointer_type, generic_recursion_map);
+  }
+  else
+  {
+    return pointer_type;
+  }
+}
+
+/// Find the concrete type assigned to a java generic parameter based on the
+/// current map of parameters -> types.
+/// Example:
+/// - generic parameter: T
+/// - map: T -> U; U -> String
+/// - result: String
+/// \param parameter java generic parameter to be specialized
+/// \param generic_recursion_map map of types for all generic parameters in
+/// the current context
+/// \return the concrete type of the parameter
+pointer_typet select_pointer_typet::find_concrete_type(
+  const pointer_typet &parameter,
+  const generic_recursion_mapt &generic_recursion_map) const
+{
+  PRECONDITION(is_java_generic_parameter(parameter));
+  const irep_idt &parameter_name =
+    to_java_generic_parameter(parameter).get_name();
+
+  if(!generic_recursion_map.count(parameter_name))
```

Comment:
Implict conversion from `int` to `bool` should be avoided IMO. I'd prefer either:

```C++
generic_recursion_map.count(parameter_name) != 0
```
or
```C++
generic_recursion_map.find(parameter_name) != generic_recursion_map.end()
```